contracts/dapis/Median.sol

```diff
@@ -21,19 +21,48 @@ contract Median is Sort, Quickselect {
             if (arrayLength % 2 == 1) {
                 return array[arrayLength / 2];
             } else {
-                return
-                    (array[arrayLength / 2 - 1] + array[arrayLength / 2]) / 2;
+                assert(arrayLength != 0);
+                unchecked {
+                    return
+                        average(
+                            array[arrayLength / 2 - 1],
+                            array[arrayLength / 2]
+                        );
+                }
             }
         } else {
             if (arrayLength % 2 == 1) {
                 return array[quickselectK(array, arrayLength / 2)];
             } else {
-                (uint256 mid1, uint256 mid2) = quickselectKPlusOne(
-                    array,
-                    arrayLength / 2 - 1
-                );
-                return (array[mid1] + array[mid2]) / 2;
+                uint256 mid1;
+                uint256 mid2;
+                unchecked {
+                    (mid1, mid2) = quickselectKPlusOne(
+                        array,
+                        arrayLength / 2 - 1
+                    );
+                }
+                return average(array[mid1], array[mid2]);
             }
         }
     }
+
+    /// @notice Averages two signed integers without overflowing
+    /// @param x Integer x
+    /// @param y Integer y
+    /// @return Average of integers x and y
+    function average(int256 x, int256 y) private pure returns (int256) {
```

Comment:
This now emulates `(x + y) / 2` with no overflow/underflow risk while only using the cheapest operations. See the tests.

| x  | y | result |
| ------------- | ------------- |  ------------- |
| 2  | 1  | 1 |
| 2  | 2  | 2 |
| 1  | 1  | 1 |
| -2  | -1  | -1 |
| -1 | -1  | -1 |
| -2  | -2  | -2 |
| 2  | -1  | 0 |
| 2  | -2  | 0 |
| 1  | -1  | 0 |
| 1  | -2  | 0 |